### PR TITLE
Fix race on enqueue query in Replicated database

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a bug in `Replicated` database engine that might rarely cause some replica to skip enqueued DDL query

### CODE REVIEW

**Comment by alesapin** [comment]
Suspicious:
```
2021-05-31 19:12:35 00910_zookeeper_test_alter_compression_codecs_long:                     [ FAIL ] 31.20 sec. - return code 159
2021-05-31 19:12:35 Received exception from server (version 21.7.1):
2021-05-31 19:12:35 Code: 159. DB::Exception: Received from localhost:9000. DB::Exception: Watching task /test/clickhouse/db/default/log/query-0000000004 is executing longer than distributed_ddl_task_timeout (=30) seconds. There are 2 unfinished hosts (2 of them are currently active), they are going to execute the query in background. 
2021-05-31 19:12:35 , result:
2021-05-31 19:12:35 
```

**Comment by tavplubix** [comment]
It fails for different reason:
```
$ zgrep -Fa "CREATE TABLE test_9otca3.alter_compression_codec1" clickhouse-server1.log.gz24805
2021.05.31 19:12:05.110716 [ 89546 ] {} <Debug> DDLWorker(test_9otca3): Executing query: CREATE TABLE test_9otca3.alter_compression_codec1 UUID 'ab2c1f37-bf7b-417e-ab2c-1f37bf7b717e' (`somedate` Date CODEC(LZ4), `id` UInt64 CODEC(NONE)) ENGINE = ReplicatedMergeTree(concat('/clickhouse/tables/test_00910/', currentDatabase(), 'alter_compression_codecs'), '1') PARTITION BY somedate ORDER BY id
2021.05.31 19:12:05.111096 [ 89546 ] {35fa99ed-b209-4e30-93be-b74b5aa0cd91} <Debug> executeQuery: (from 0.0.0.0:0, user: , using production parser) /* ddl_entry=query-0000000004 */ CREATE TABLE test_9otca3.alter_compression_codec1 UUID 'ab2c1f37-bf7b-417e-ab2c-1f37bf7b717e' (`somedate` Date CODEC(LZ4), `id` UInt64 CODEC(NONE)) ENGINE = ReplicatedMergeTree(concat('/clickhouse/tables/test_00910/', currentDatabase(), 'alter_compression_codecs'), '1') PARTITION BY somedate ORDER BY id
2021.05.31 19:12:05.160707 [ 89546 ] {35fa99ed-b209-4e30-93be-b74b5aa0cd91} <Error> executeQuery: Code: 253, e.displayText() = DB::Exception: Replica /clickhouse/tables/test_00910/test_9otca3alter_compression_codecs/replicas/1 already exists. (version 21.7.1.7003) (from 0.0.0.0:0) (in query: /* ddl_entry=query-0000000004 */ CREATE TABLE test_9otca3.alter_compression_codec1 UUID 'ab2c1f37-bf7b-417e-ab2c-1f37bf7b717e' (`somedate` Date CODEC(LZ4), `id` UInt64 CODEC(NONE)) ENGINE = ReplicatedMergeTree(concat('/clickhouse/tables/test_00910/', currentDatabase(), 'alter_compression_codecs'), '1') PARTITION BY somedate ORDER BY id), Stack trace (when copying this message, always include the lines below):
```
Path and replica name are the same on all hosts:
`ReplicatedMergeTree(concat('/clickhouse/tables/test_00910/', currentDatabase(), 'alter_compression_codecs'), '1')`

So test was broken

**Comment by tavplubix** [comment]
Functional stateless tests flaky check - OK
Integration tests (release) - test_distributed_queries_stress - `2021.06.02 01:49:17.910517 [ 1 ] {} <Fatal> Application: Child process was terminated by signal 11.`, probably it's related to libunwind
Performance - `2021.06.02 12:30:59.488208 [ 175 ] {} <Fatal> Application: Child process was terminated by signal 11.`, same
SQLancer test - `Table default.t0 already exists` - looks like SQLancer works incorrectly, cc: @qoega 
Yandex third-party checks - no related failures

File: src/Databases/DatabaseReplicatedWorker.cpp [hunk 1]
```diff
@@ -72,25 +73,63 @@ void DatabaseReplicatedDDLWorker::initializeReplication()
...
+    const String query_path_prefix = database->zookeeper_path + "/log/query-";
 
     /// We cannot create sequential node and it's ephemeral child in a single transaction, so allocate sequential number another way
     String counter_prefix = database->zookeeper_path + "/counter/cnt-";
-    String counter_path = zookeeper->create(counter_prefix, "", zkutil::CreateMode::EphemeralSequential);
+    String counter_lock_path = database->zookeeper_path + "/counter_lock";
+
+    String counter_path;
+    size_t iters = 1000;
+    while (--iters)
```
Comment (alesapin) [inline src/Databases/DatabaseReplicatedWorker.cpp:90]: Maybe add sleep, but this is also OK.

**Review by alesapin — APPROVED** [review]

**Review by alesapin — APPROVED** [review]

**Pull request merged**